gstreamer 1.4 -> 1.6 and pitivi 0.94 -> 0.95

bjornfor opened the pull request:
The plan was just to update pitivi, but as it depends on gstreamer >= 1.6.0 I did both in one go. Also gst-validate had to be added, and pygobject3 needed to be updated for pitivi to be happy.

I've tested pitivi pretty well (so I know it doesn't regress), but I haven't tested any other gstreamer packages yet. (GStreamer documentation says 1.x is compatible/stable, but I don't know...)

CC @iyzsong (gstreamer _and_ pitivi maintainer)

DON'T MERGE YET, SEVERE GSTREAMER PERFORMANCE REGRESSION.

bjornfor committed:
pitivi: 0.94 -> 0.95

Notable (hard) dependency changes:
  - gstreamer >= 1.6.0
  - gst-plugins-bad with GTK support (need "gstgtk" plugin)
  - pygobject >= 3.14
  - gst-validate
  - matplotlib (used for drawing the "timeline")

Changes v1 -> v2:
pitivi no longer requires clutter, so I switched clutter-gtk => gtk3.

bjornfor committed:
gstreamer: 1.4.x -> 1.6.x (all modules)

(And while at it, gst-vaapi 0.6.0 -> 0.6.1.)

* gst-editing-services grew additional build time dependencies, flex and
  perl.

* gst-libav switched from libav to ffmpeg as "libav" provider, see
  http://gstreamer.freedesktop.org/releases/1.6/.
  Without using ffmpeg, one may hit issues such as this (which I
  initially did):

  (gst-plugin-scanner:19751): GStreamer-WARNING **: Failed to load plugin '/nix/store/0wgpq2yx9wrkp2mh4rn1c7zbiq2bqa2l-gst-libav-1.6.1/lib/gstreamer-1.0/libgstlibav.so':
  /nix/store/0wgpq2yx9wrkp2mh4rn1c7zbiq2bqa2l-gst-libav-1.6.1/lib/gstreamer-1.0/libgstlibav.so: undefined symbol: av_frame_get_sample_rate

bjornfor committed:
gst-validate: init at 1.6.0

pitivi-0.95 requires this. (I added iyzsong as maintainer, in line with
the other gstreamer packages.)

bjornfor committed:
pygobject3: 3.12.1 -> 3.18.2

pitivi-0.95 needs pygobject >= 3.14.

bjornfor commented:
I'm planning to test my whole system with this PR. If I don't hit any issues, I'll remove the "WIP" label.

bjornfor commented:
Uhm, well, to say that pitivi 0.95 doesn't regress... I mean in respect to packaging at least :-)  It's supposed to be a lot more stable, but I'm hitting some issues that were in 0.94. And they have removed undo/redo functionality (mentioned here: http://jeff.ecchi.ca/blog/2015/11/19/pitivi-0-95-enfant-suisse/). It'll be re-added when it's deemed stable.

vcunat commented:
It would be good to try running some (other) gst stuff, preferably before merging.

iyzsong commented:
Look good to me, thanks!
I don't see the libgstlibav.so warnning.
Test with parole, and it work fine.

OK to merge?

bjornfor commented:
The libgstlibav.so warning happened before I switched libav to ffmpeg. It doesn't happen now.

I rebuilt my system and tested cheese (capturing video with webcam). It sort of worked, but was really slow and dropped frames / audio. I don't know if it was like this before, as I'm now on my laptop (my main PC crashed). I'll test some more (back and forth).

I noticed in the release notes that pitivi no longer needs "clutter", so I removed that from buildInputs.

bjornfor commented:
After testing cheese with/without this PR, it seems to regress somewhat. It seems to struggle with keeping up with the data rate:

```
(cheese:5562): cheese-WARNING **: Can't record audio fast enough: gstaudiobasesrc.c(866): gst_audio_base_src_create (): /GstCameraBin:camerabin/GstAutoAudioSrc:audiosrc/GstPulseSrc:audiosrc-actual-src-puls:
Dropped 10584 samples. This is most likely because downstream can't keep up and is consuming samples too slowly.
```

On master branch (eeb2935ac5a74eeb3d2e32f311fd6e6ef743e78c) I can capture 1280x720 without any warnings / dropped samples. But with this PR, I have to reduce the resolution to 320x240 (or else the recording will have lots of dropped video/audio samples => really poor video).

bjornfor commented:
I was thinking this was due to the switch from libav to ffmpeg in gst-libav, but cheese doesn't use gst-libav. So this is simply a really bad performance regression in gstreamer!?

bjornfor commented:
A gstreamer pipeline for testing:

```
gst-launch-1.0 -e webmmux name=mux ! filesink location=test.webm v4l2src ! video/x-raw,width=1280,height=720 ! videoconvert ! vp8enc ! queue ! mux. autoaudiosrc ! audio/x-raw,rate=44100 ! vorbisenc ! queue ! mux.
```

And a complete command line, with nix-shell wrapping:

```
nix-shell -p gst_all_1.gstreamer gst_all_1.gst-plugins-base gst_all_1.gst-plugins-good gst_all_1.gst-plugins-bad gst_all_1.gst-plugins-ugly less -I nixpkgs=. --run "gst-launch-1.0 -e webmmux name=mux ! filesink location=test.webm v4l2src ! video/x-raw,width=1280,height=720 ! videoconvert ! vp8enc ! queue ! mux. autoaudiosrc ! audio/x-raw,rate=44100 ! vorbisenc ! queue ! mux."
```

The above works just fine in gstreamer 1.4, but has performance issues in gstreamer 1.6 ("Dropped N samples" after a few seconds).

What do do? Ignore it and merge? I don't know what we as packagers (downstreams) can do here...

iyzsong commented:
Oh, this is bad, 
I know little about recording and encoding, would you like to report it to upstream?

bjornfor commented:
@iyzsong: I can give it a try.

bjornfor commented:
Upstream bug report: https://bugzilla.gnome.org/show_bug.cgi?id=759141

bjornfor commented:
Would anyone mind trying to reproduce this? I got thiblahute in pitivi IRC channel to reproduce it (so it's not just my system), but the question is whether this is significant enough to hold off the gstreamer upgrade.

This issue happens on my system when the CPU load is at 100%, so it might just be a question of "how graceful gstreamer can be at reducing the recording quality when overloaded". I.e. not directly a performance thing.

bjornfor commented:
Ok, we're getting somewhere on the upstream bug: https://bugzilla.gnome.org/show_bug.cgi?id=759141

bjornfor commented:
Summary from upstream bug:

It seems the default behaviour for the vp8enc element is to produce "best quality". This hogs the CPU and starves the audio element until it drops samples. It's trivial to fix on the command line by specifying "vp8enc deadline=1" (which tells the encoder that the frame has a 1 us deadline, a.k.a. real-time mode).

I'm not sure where the differing behaviour comes from (gst 1.4 -> 1.6), but I really think Cheese should specify "deadline=1" in its gstreamer pipeline. That is at least one way to fix the issue.

iyzsong commented:
Thanks for the report and detailed explanation!

It turn out that cheese did  specify "deadline=1" as defaults in:
```
  $gst-plugins-good/share/gstreamer-1.0/presets/GstVP8Enc.prs
``` 
but it can't find the vp8enc preset in gst-plugins-good, complain with:
```
(cheese:9615): cheese-WARNING **: Can't find vp8enc preset: "Profile Realtime",
using alternate preset: "Cheese Realtime".
If you see this, make a bug report!
```
then `~/.local/share/gstreamer-1.0/presets/GstVP8Enc.prs` is used, which has `deadline=0`.

Dose cheese works fine with "Profile Realtime"?
(copy GStVP8Enc.prs from gst-plugins-good to `~/.local/share/gstreamer-1.0/presets/`)

bjornfor commented:
Oh, I missed that warning (didn't pay enough attention)! Thanks!

Right, copying ```$gst-plugins-good/share/gstreamer-1.0/presets/GstVP8Enc.prs``` to ```~/.local/share/gstreamer-1.0/presets/GstVP8Enc.prs``` fixes it (no audio drop-outs). It is also possible to just edit ```~/.local/share/gstreamer-1.0/presets/GstVP8Enc.prs``` in-place (after Cheese has made the "Cheese Realtime" preset there): deadline=0 -> deadline=1.

* Why doesn't cheese find the "Profile Realtime" preset in gst-plugins-good?
* Why does it create a "Cheese Realtime" preset with default=0? It knows very well that "realtime" means "default=1" :-)

vcunat commented:
Right, so except for cheese, gst-1.6 seems fine so far?

bjornfor commented:
AFAIK, yes.

bjornfor commented:
Well, one could argue that it is a regression that the vp8enc element starves the audio element in 1.6 and not nearly as much in 1.4. But that's insignificant, IMHO. I think we should merge and then fix Cheese (make it find the proper preset file).

Possibly gstreamer will provide a better default config in the future: https://bugzilla.gnome.org/show_bug.cgi?id=759141#c8.

iyzsong commented:
I open a new issue for cheese, OK to merge this one?

vcunat commented:
Yes, it sounds fine to me.
